File: 프로그래머스/연습문제/Lv2/뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

vector<int> solution(vector<int> numbers) {
    vector<int> answer(numbers.size(), -1);
    stack<int> st;
    st.push(numbers[numbers.size()-1]);
    
    for(int i=numbers.size()-2; i>=0; i--){
        while(st.size() && (st.top()<=numbers[i]))
            st.pop();
        
        if(st.size()) answer[i] = st.top();
        st.push(numbers[i]);
    }
    
    return answer;
}

File: 프로그래머스/연습문제/Lv2/뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [-1 for i in range(len(numbers))]
    st = [numbers[-1]]

    for i in range(len(numbers)-2, -1, -1):
        while len(st)>0 and st[-1]<=numbers[i]:
            st.pop(-1)

        if len(st) > 0: answer[i] = st[-1]
        st.append(numbers[i])

    return answer
